Make remoteCacheId content-based

eed3si9n opened the pull request:
Fixes #5842

Using Git commit for remote caching id was one of the first things people think about, but it doesn't scale because that means you're downloading all module JARs all the time anytime anyone adds a commit. The basis of remote caching is that people are adding commits to many subprojects, so that's bad.

As an alternative, this uses content hash of input files and external dependency JARs. `remoteCacheId`, `pushRemoteCache`, and `pullRemoteCache` are now scoped to configuration (`Compile` and `Test`). Since we're not resolving any version conflicts, dependency resolver is no longer used and the artifacts are directly downloaded from the Ivy resolver. This should speed up the `pullRemoteCache` since it avoids the POM download as well.

### note

For sbt-binrary-remote-cache this created a bit of complication since the (publishing) resolver doesn't act correctly as (downloading) resolver in terms of the credentials, so I had to create a new key `remoteCacheResolvers` to have asymmetric resolver.

eed3si9n committed:
1.4.2-SNAPSHOT

eed3si9n committed:
Make remoteCacheId content-based

Fixes https://github.com/sbt/sbt/issues/5842

bjaglin commented on main/src/main/scala/sbt/internal/RemoteCache.scala:
```diff
@@ -41,12 +45,30 @@ object RemoteCache {
+      val compileExtraInc = (Compile / extraIncOptions).value
+      val compileInputs = (Compile / unmanagedSources / inputFileStamps).value
+      val compileCp = (Compile / externalDependencyClasspath / outputFileStamps).value
+      val testExtraInc = (Compile / extraIncOptions).value
```
```suggestion
      val testExtraInc = (Test / extraIncOptions).value
```

eed3si9n replied:
Thanks!

bjaglin commented on main/src/main/scala/sbt/internal/RemoteCache.scala:
```diff
@@ -41,12 +45,30 @@ object RemoteCache {
   )
 
   lazy val projectSettings: Seq[Def.Setting[_]] = (Seq(
+    remoteCacheId := {
+      val compileExtraInc = (Compile / extraIncOptions).value
+      val compileInputs = (Compile / unmanagedSources / inputFileStamps).value
+      val compileCp = (Compile / externalDependencyClasspath / outputFileStamps).value
```
Could this be defined at the configuration level (and combined at the project level via a scope filter), so that other configurations such as IntegrationTest can more easily be supported? https://github.com/sbt/sbt/blob/5a960f2f94b8c5260e07e380b76e88f1d6959cd4/main/src/main/scala/sbt/internal/RemoteCache.scala#L206-L207 would still need to be exposed, maybe within the same `Seq[Def.Setting[_]]` ?

eed3si9n replied:
Good idea. I'm hoping I can eventually implement `pullRemoteCache` at the configuration level.

eed3si9n committed:
Refactor remoteCacheId

eed3si9n committed:
Make sure we test the caching effect

eed3si9n committed:
Refactor remote cache to per-config

Refactor remote caching to be scoped to configuration.
In addition, this avoid the use of dependency resolver (since I'm not resolving anything) and directly invoke the Ivy resolver for the artifact, somewhat analogus to publishing process.
This should speed up the `pullRemoteCache` since it avoids the POM download as well.

For sbt-binrary-remote-cache this created a bit of complication since the (publishing) resolver doesn't act correctly as (downloading) resolver in terms of the credentials, so I had to create a new key `remoteCacheResolvers` to have asymmetric resolver.

eed3si9n committed:
Bring back fixed-id for testing
